src/lnurl.js

```diff
@@ -0,0 +1,116 @@
+import bech32 from 'bech32'
+import wrap from './lib/promise-wrap'
+
+const debug = require('debug')('lightning-charge')
+
+module.exports = (app, payListen, model, auth) => {
+  const {
+    newInvoice, listInvoicesByLnurlPayEndpoint
+  , getLnurlPayEndpoint, listLnurlPayEndpoints
+  , setLnurlPayEndpoint, delLnurlPayEndpoint
+  } = model
+
+  app.get('/endpoints', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      (await listLnurlPayEndpoints())
+        .map(lnurlpay => addBech23Lnurl(req, lnurlpay))
+    )))
+
+  app.post('/endpoint', auth, wrap(async (req, res) =>
+    res.status(201).send(
+      addBech23Lnurl(req, await setLnurlPayEndpoint(null, req.body))
+    )))
+
+  app.put('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      addBech23Lnurl(req, await setLnurlPayEndpoint(req.params.id, req.body))
+    )))
+
+  app.delete('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(await delLnurlPayEndpoint(req.params.id))))
+
+  app.get('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      addBech23Lnurl(req, await getLnurlPayEndpoint(req.params.id))
+    )))
+
+  app.get('/endpoint/:id/invoices', auth, wrap(async (req, res) =>
+    res.send(await listInvoicesByLnurlPayEndpoint(req.params.id))))
+
+  // this is the actual endpoint users will hit
+  app.get('/lnurl/:id', wrap(async (req, res) => {
+    const lnurlpay = await getLnurlPayEndpoint(req.params.id)
+
+    res.status(200).send({
+      tag: 'payRequest'
+    , minSendable: lnurlpay.min
+    , maxSendable: lnurlpay.max
+    , metadata: makeMetadata(lnurlpay)
+    , commentAllowed: lnurlpay.comment
+    , callback: `https://${req.hostname}/lnurl/${lnurlpay.id}/callback`
+    })
+  }))
+
+  app.get('/lnurl/:id/callback', wrap(async (req, res) => {
+    const lnurlpay = await getLnurlPayEndpoint(req.params.id)
+
+    if (req.query.amount > lnurlpay.max)
+      return res.send({status: 'ERROR', reason: 'amount too large'})
+    if (req.query.amount < lnurlpay.min)
+      return res.send({status: 'ERROR', reason: 'amount too small'})
+
+    let invoiceMetadata = {...req.query}
+    delete invoiceMetadata.amount
+    delete invoiceMetadata.fromnodes
+    delete invoiceMetadata.nonce
+    invoiceMetadata = {...lnurlpay.metadata, ...invoiceMetadata}
+
+    const invoice = await newInvoice({
+      descriptionHash: require('crypto')
+        .createHash('sha256')
+        .update(makeMetadata(lnurlpay))
+        .digest('hex')
+    , msatoshi: req.query.amount
+    , metadata: invoiceMetadata
+    , webhook: lnurlpay.webhook
+    , lnurlpay_endpoint: lnurlpay.id
+    })
+
+    let successAction
+    if (lnurlpay.success_url) {
+      successAction = {
+        tag: 'url'
+      , url: lnurlpay.success_url
+      , description: lnurlpay.success_text || ''
+      }
+    } else if (lnurlpay.success_value) {
```

Comment:
I'm assuming this should be `success_secret`?

Are you planning to implement this as part of this PR, or maybe later in a followup one? If later, I would remove this and the field from the model definition for now.

Reply:
I'm planning to implement it as part of this, hopefully.

Reply:
I realized it doesn't make much sense to implement this for now as all use cases I can imagine for it will need a different `success_secret` for each payment and I don't know how was that supposed to happen in this context, so I guess I'll remove this and leave it for a potential future.